dist/steps/run_tests.sh

```diff
@@ -169,6 +205,44 @@ if [ "$PLAY_MODE" = "true" ]; then
   cat "$FULL_ARTIFACTS_PATH/playmode-results.xml" | grep test-run | grep Passed
 fi
 
+#
+# Combine test results if needed
+#
+COMBINE_EXIT_CODE=0
+if [ "$EDIT_MODE" = "true" ] && [ "$PLAY_MODE" = "true" ] && [ "$ENABLE_CODE_COVERAGE" = "true" ]; then
+  echo ""
+  echo "##############################"
+  echo "# Combining Coverage Results #"
+  echo "##############################"
+  echo ""
+  unity-editor \
+    -batchmode \
+    -debugCodeOptimization \
+    -enableCodeCoverage \
+    -logFile "$FULL_ARTIFACTS_PATH/combine_coverage.log" \
+    -projectPath "$UNITY_PROJECT_PATH" \
+    -coverageResultsPath "$FULL_COVERAGE_RESULTS_PATH" \
+    -coverageOptions "generateHtmlReport;generateBadgeReport" \
+    -quit
+
+  # Catch exit code
+  COMBINE_EXIT_CODE=$?
+
+  # Print unity log output
+  cat "$FULL_ARTIFACTS_PATH/combine_coverage.log"
+
+  # Display results
+  if [ $COMBINE_EXIT_CODE -eq 0 ]; then
+    echo "Combine coverage succeeded, no failures occurred";
+  elif [ $COMBINE_EXIT_CODE -eq 2 ]; then
+    echo "Combine coverage, some tests failed";
+  elif [ $COMBINE_EXIT_CODE -eq 3 ]; then
+    echo "Combine coverage (other failure)";
+  else
+    echo "Unexpected exit code $COMBINE_EXIT_CODE";
+  fi
+fi
+
```

Comment:
I'd like to think a bit more before committing to having another duplicated block in bash.

Ideally we would not need 3 separate blocks of code just to run the editor for playMode, editMode, or both.

Reply:
Given the previous comment, I might adjust this so that we have run commands. And if the player selected "All" as the test platform, then in the javascript code, we could generate the three blocks of parameters to run (for edit mode, playmode, and then the third one to combine them).

Reply:
Ok, I wrote some adjustments to the `run_tests.sh` file. We could determine most of those parameters from the javascript however, but we need to change the logic to run the unity program a variable number of times. I achieved this by making the `TEST_PLATFORMS` parameter which will have a list of test platforms separated by a `;` with a special platform `COMBINE_RESULTS` for combining the results of previous tests. 

If we really don't want to make any changes to the bash script, then this change can be avoided. Just an idea of how we could implement this differently.

Reply:
Long term we intend to create a plugin system. We could solve exactly this problem using a plugin system that people opt-in to, which would execute some additional code. The code would be conditionally injected into bash (by javascript) using a variable.

Even then we could still keep the loop to iterate over PlayMode and EditMode.

Going with your current approach:

Right now because of the following code it would never combine the results in case I were to choose PlayMode only.

```ts
    const testPlatforms = (
      testMode === 'all' ? ['playmode', 'editmode', 'COMBINE_RESULTS'] : [testMode]
    ).join(';');
```

But does that mean the report is still generated? Or does the generation solely happen inside that combine results step?

A) If generation would work like this then I'd say lets go with this approach. 

B) If it does not work and it requires additional logic, perhaps we can inject the whole command as if it were a plugin from TS into the shell script using a variable already.

Our plugins system at its simplest would look roughly like this (consider psuedo) (I didn't get to implementing it yet)

```ts
// typescript:

  const runEditor = (args) => exec(`
   unity-editor \
     -batchmode \
    -logFile "$FULL_ARTIFACTS_PATH/$platform.log" \
    -projectPath "$UNITY_PROJECT_PATH" \
   ${args}`


  class Plugin { 
    static editorCall() {
      runEditor(`
        -coverageResultsPath "$FULL_COVERAGE_RESULTS_PATH" \
        -enableCodeCoverage \
        -debugCodeOptimization \
        -coverageOptions "$COVERAGE_OPTIONS" \
        // omitted CUSTOM_PARAMETERS, as i'ts only a combine step that's internal to the plugin itself
      `)
    }
  }

  // somewhere after docker.run()
 for(const plugin of pluginsRegistry) {
  plugin.postRuntimeCalls?.()
 }
```

and omit it from bash entirely (or even leave the bash script untouched entirely).

Now I don't expect you to implement all that, but I'd love your thoughts on this, as well as on what would be the best intermediate step in case (A) is not applicable.

Reply:
Actually option (A) is how it works. Right now, since we use the default `-coverageOptions` as provide it would work like this (when coverage options = `enableCyclomaticComplexity;generateHtmlReport;generateBadgeReport`)

When running just `playmode` or `editmode`, it will generate an html report and coverage results

When running `all` it will do the following steps
1. Generate a html report and coverage results for `playmode`.
2. Overwrite the previous html report for `playmode` and make a new coverage results for only `editmode`. 
3. The `COMBINE_RESULTS` run will overwrite the previous html report for `editmode` and create a new combined report from the results of `editmode` and `playmode` as well as a combined coverage file. 

The flags `generateHtmlReport` and `generateBadgeReport` tell the code coverage program to generate and overwrite the existing report with the results of the current test. since when we use `COMBINE_REUSLTS` we only have the `-quit` and `-testResults` flags, unity knows to combine all the current results in the output folder. You can see an example of this in the test runs. I'll include the code coverage reports in the `main.yml` example file for running in each type of mode actually (just upload the `CodeCoverage` output folder as well). 

Also, unity supports running more test profiles than just `playmode` or `editmode` if the user defines custom ones so hopefully this could allow for more user customization if the end user wants to in the future. 

I think adding a `Plugin` run in the future would probably be a better solution than the current `COMBINE_REUSLTS` but we can leave that to a future PR. Maybe instead of injecting the flags for what modes to run via environment variables, we could create an XML or variables file, pass it into the container, and change the bash script to iterate through each element in the XML file and read the customer arguments based on the values in the XML. would be easy to parse and run, easy to add more customization and debugging? 

I can open up an issue to start investigating how to structure those XML files and come up with a basic format for these runs. It could even be extended to the other plugins like the `unity-builder` if we would ever need to run multiple builds as part of one run.